### ISSUE TITLE
mechanicalsoup does not respect order of inputs in a form

### ISSUE THREAD

**Issue by arabidopsis** [MechanicalSoup/MechanicalSoup#244]
First... thanks for all your hard work; much appreciated.

mechanical soup does not respect order of inputs in a form. This is because BeautifulSoup
does not either. I think you need to replace in `Browser._request `  the line:

`for tag in form.select(selector):`

with (maybe!):
      
`for _, tag in sorted((form.index(tag), tag) for tag in form.select(selector)):`

Yes there seem to be websites where order matters (who knew?!)

**Comment by hemberger** [comment]
Hi, and thanks for reporting this issue!

I think there are two questions here:
1. What is the standard for encoding form data in a query string?
2. What is the standard for parsing the query string?

Below are some potentially useful references.

[Wikipedia: URL Normalization](https://en.wikipedia.org/wiki/URL_normalization#Normalizations_that_change_semantics):
> However, the order of parameters in a URL may be significant (this is not defined by the standard)

[Wikipedia: Query String](https://en.wikipedia.org/wiki/Query_string#Structure):
> The order of the queries doesn't matter (name=ferret&color=purple and color=purple&name=ferret both produce the same results) and the exact structure of the query string is not standardized.

The [URI RFC](https://tools.ietf.org/html/rfc3986#section-3.4) seems to corroborate that there is no standard for parsing the query string.

Regarding form content encoding, [W3C](https://www.w3.org/TR/html401/interact/forms.html#h-17.13.4.1) seems to be a bit more explicit about it:
> The control names/values are listed in the order they appear in the document. The name is separated from the value by '=' and name/value pairs are separated from each other by '&'.

The `bs4.Tag.element.select` method that we use in `Browser._request`seems to determine the order of results based on the order of comma-delimited selectors. Our selector is `[input, button, textarea, select]`, and it looks like they are correctly ordered within each of those 4 categories, but not correct overall.

**tl;dr**: I think you are right, and we should probably find a way to order them correctly. Your suggestion seems sensible, but someone will need to write a test case. I'll do it when I can if no one else volunteers. ;)

**Comment by noahbarnette** [comment]
@arabidopsis This hasn't really been an issue for me but have you considered offering a fix and committing it? If you have a specific use case where this is a big issue, you may be more knowledgeable as to how it should best be implemented than others.

**Comment by hemberger** [comment]
Just an update on this issue: if you update to BeautifulSoup 4.7.0+, this issue will be fixed (i.e. the request will be page-ordered). See #257 for more details.

I wonder if we should make BeautifulSoup 4.7.0 the minimum requirement for MechanicalSoup?
